resolve: follow-ups for extra DNS stub listener

yuwata opened the pull request:
Follow-ups for #16044.

Closes #16948.

poettering commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -460,12 +460,26 @@ static int manager_dns_stub_udp_fd(Manager *m) {
+        if (l->family == AF_INET)
+                sa = (union sockaddr_union) {
+                        .in.sin_family = l->family,
+                        .in.sin_port = l->port != 0 ? htobe16(l->port) : htobe16(53),
```
doesn't matter, but you could move the ternary op inside the htobe16()? (might need to suffix 53 with U then though)

poettering commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -479,7 +493,7 @@ static int manager_dns_stub_udp_fd_extra(Manager *m, DNSStubListenerExtra *l) {
                 goto fail;
 
-        if (bind(fd, &l->address.sockaddr.sa, l->address.size) < 0) {
+        if (bind(fd, &sa.sa, l->family == AF_INET ? sizeof(struct sockaddr_in) : sizeof(struct sockaddr_in6)) < 0) {
```
maybe we should add a helper for this one day: 

SOCKADDR_LEN() which checks the family and then returns sizeof(sockaddr_in), sizeof(sockaddr_in6) or SOCKADDR_UN_LEN() or SOCKADDR_LL_LEN() or -EAFNOSUPPORT or so.

poettering commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -413,11 +413,17 @@ static int set_dns_stub_common_socket_options(int fd) {
+                r = setsockopt_int(fd, IPPROTO_IP, IP_RECVTTL, true);
+                if (r < 0)
+                        return r;
+        }
```
hmm, but we should set the matching optoins for ipv6 here, i.e. IPV6_RECVPKTINFO + IPV6_RECVHOPLIMIT

poettering commented on src/resolve/resolved-dns-stub.c:
```diff
@@ -485,11 +491,13 @@ static int manager_dns_stub_udp_fd_extra(Manager *m, DNSStubListenerExtra *l) {
-        if (r < 0)
-                goto fail;
+        if (l->family == AF_INET) {
+                r = setsockopt_int(fd, IPPROTO_IP, IP_FREEBIND, true);
```
should call IPV6_FREEBIND socktopt on ipv6

keszybz commented on man/resolved.conf.xml:
```diff
@@ -271,27 +271,27 @@
+        <literal>:</literal>. If the protocol is not specified, the service will listen on both UDP and
+        TCP. It may be also optionally suffixed by a numeric port number with separator
+        <literal>:</literal>. When an IPv6 address is specified with a port number, then the address
+        must be in the square brackets. If the port is not specified, then the service takes port as 53.
```
"takes port as 53" doesn't make much sense. Just say "uses port 53"

keszybz reviewed (commented):
This partially overlaps with #16947. Let's merge yours first, and then I'll rebase my changes on top.

yuwata commented:
@poettering and @keszybz Thank you for the review. I've force-pushed a revised version, which addresses all your comments, and fixes several critical bugs.

keszybz commented:
bionic-i386: 347/811 test-process-util                       TIMEOUT 30.06 s , looks unrelated.

keszybz commented:
LGTM.
